ecosystem/typescript/sdk/src/transactions/transaction_worker.ts

```diff
@@ -0,0 +1,225 @@
+   * @param sentTransaction transactions that were sent to chain and are now waiting to be executed
+   * @param sequenceNumber the account's sequence number that was sent with the transaction
+   */
+  async checkTransaction(sentTransaction: PromiseFulfilledResult<PendingTransaction>, sequenceNumber: bigint) {
+    const waitFor: Array<Promise<Transaction>> = [];
+    waitFor.push(this.provider.waitForTransactionWithResult(sentTransaction.value.hash, { checkSuccess: true }));
+    const sentTransactions = await Promise.allSettled(waitFor);
+
+    for (let i = 0; i < sentTransactions.length; i += 1) {
+      const executedTransaction = sentTransactions[i];
+      if (executedTransaction.status === "fulfilled") {
+        // transaction executed to chain
+        this.executedTransactions.push([executedTransaction.value.hash, sequenceNumber, null]);
+        this.emit(transactionExecuted, [this.executedTransactions.length, executedTransaction.value.hash]);
+      } else {
+        // transaction execution failed
+        this.executedTransactions.push([executedTransaction.status, sequenceNumber, executedTransaction.reason]);
+        this.emit(executionFailed, [this.executedTransactions.length, executedTransaction.reason]);
+      }
+    }
+  }
+
+  /**
+   * Push transaction to the transactions queue
+   * @param payload Transaction payload
+   */
+  async push(payload: TxnBuilderTypes.TransactionPayload): Promise<void> {
+    await this.transactionsQueue.push(payload);
+  }
+
+  /**
+   * Generates a signed transaction that can be submitted to chain
+   * @param account an Aptos account
+   * @param sequenceNumber a sequence number the transaction will be generated with
+   * @returns
+   */
+  async generateNextTransaction(account: AptosAccount, sequenceNumber: bigint): Promise<Uint8Array | undefined> {
+    if (this.transactionsQueue.length === 0) return undefined;
+    const payload = await this.transactionsQueue.shift()!;
+    const rawTransaction = await this.provider.generateRawTransaction(account.address(), payload, {
+      providedSequenceNumber: sequenceNumber,
+    });
+    const signedTransaction = await AptosClient.generateBCSTransaction(account, rawTransaction);
+    return signedTransaction;
+  }
+
+  /**
+   * Starts transaction submission and transaction processing.
+   */
+  async run() {
+    try {
+      while (!this.stopped) {
+        /* eslint-disable no-await-in-loop, no-promise-executor-return */
+        await Promise.all([this.submitNextTransaction(), this.processTransactions()]);
+        /** 
+         * since run() function runs continuously in a loop, it prevents the execution 
+         * from reaching a callback function (e.g when client wants to gracefuly stop the worker). 
+         * Add a small delay between iterations to allow other code to run
+        /* eslint-disable no-await-in-loop */
+        await new Promise((resolve) => setTimeout(resolve, 100));
```

Comment:
the reason for having the delay is that the `run` method continuously calls `submitNextTransaction` and `processTransactions`. This can potentially lead to blocking the event loop and preventing other asynchronous operations.

Reply:
No, JS does not provide a built-in queue implementation this is why we use Array, and having it async is with the use of Promise - which is what I thought I was doing here https://github.com/aptos-labs/aptos-core/pull/8854/files#diff-52d3c5103be2ebfaf76511751a50a8fc3ce5f720a77ddd6fed4e74891074edb0R46
I'll look at it more